docs/rendering.rst

```diff
@@ -0,0 +1,783 @@
+in-progress document to editing, with a smaller list of environment variables.
+
+source
+   *revision content* or *re-edit content*, with CKEditor parsing, passed
+   through KumaScript
+output
+   HTML content at ``/<locale>/docs/preview-wiki-content``
+database
+   *not stored*
+code
+   *not available*
+
+The *preview content* for the simple document is:
+
+.. code-block:: html
+
+   <p></p>
+
+   <p>I am a <strong>simple document</strong> with a CSS sidebar.</p>
+
+   <p style="color: red;">I am red.</p>
+
+   <h2>Some Links</h2>
+
+   <ul>
+    <li><a href="/en-US/docs/Web/HTML">The HTML Reference</a></li>
+    <li><a href="/en-US/docs/Web/HTML/Element/div" title="The HTML Content Division element (&lt;div>) is the generic container for flow content. It has no effect on the content or layout until styled using CSS."><code>&lt;div&gt;</code></a></li>
+    <li><a href="/en-US/docs/NewDocument">A new document</a></li>
+   </ul>
+
+   <div class="button"></div>
+
+   &lt;script&gt;
+     alert('How about this?');
+   &lt;/script&gt;
+
+The environment in preview is different than in regular KumaScript rendering:
+
+url
+   The base URL of the website, like ``https://developer.mozilla.org/``
+locale
+   The locale of the request, like ``"en-US"``
+
+Some macros use the absence of an environment variable to detect preview mode,
+and change their output. For example, ``{{CSSRef}}`` notices that ``env.slug``
+is not defined, and outputs an empty string, leaving ``<p></p>`` in the
+preview output.
+
+Other macros don't have specific code to detect preview mode, and have
+kumascript rendering errors.
+
+Some macros, like ``{{HTMLElement}}``, can work as expected in preview.
+
+Raw content
+===========
+A ``?raw`` parameter can be added to the end of a document to request the
+source for a revision.
+
+source
+   *cleaned content*, with filters
+output
+   The page with a ``?raw`` query parameter
+database
+   *not stored*
+code
+   *not available*
+
+For the simple document, this is the *raw content*:
+
+.. code-block:: html
+
+   <p>{{ CSSRef }}</p>
+
+   <p>I am a <strong>simple document</strong> with a CSS sidebar.</p>
+
+   <p style="color: red;">I am red.</p>
+
+   <h2 id="Some_Links">Some Links</h2>
+
+   <ul>
+    <li><a href="/en-US/docs/Web/HTML">The HTML Reference</a></li>
+    <li>{{HTMLElement('div')}}</li>
+    <li><a href="/en-US/docs/NewDocument">A new document</a></li>
+   </ul>
+
+   <div class="button"></div>
+
+   &lt;script&gt;
+     alert('How about this?');
+   &lt;/script&gt;
+
+The *cleaned content* is parsed for filtering . The headers get IDs, based on
+the content, if they did not have them before.  For example,
+``id="Some_Links"`` is added to the ``<h2>``.
+
+A simple filter is applied that strips any attributes that start with
+``on``, such as the scripting attempt ``onclick``. However, none of these
+should remain in the *cleaned content*.
+
+Live sample
```

Comment:
Can you mention how live samples are related to interactive examples and the codepen interactive examples? Is this an outdated thing, or are we still using it? It might be helpful to have a link to a document that includes one of these in addition to the link to the docs.

I also don't really get what stage of the rendering process these are handled in. What prevents the live samples <pre> sections from showing up in the body html? (And would this section be better if it was right after the body html and quick links html section?)

Reply:
The [Live samples](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Structures/Live_samples) page includes a [demo](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Structures/Live_samples#Live_sample_demo), which I'll link to.  We're still using it, and you can look for ``EmbedLiveSample`` on the [macro dashboard](https://developer.mozilla.org/en-US/dashboards/macros) to find usage (1171 English pages).

The ``<pre>`` sections often show up in the body HTML. The goal is usually to say "Here's some CSS, and here's what it looks like".

I'm currently splitting the document between the "main line" rendering sections and the alternate rendering sections. Let's see if that still makes sense once the whole thing is published.

I'll add all this to the doc, along with a quick list of the other sample types.